### PR TITLE
Metric distortion

### PR DESCRIPTION
**Pull request by MatPont**
This PR adds the MetricDistortion filter that compute information about curvature, distance and area over a surface.

It takes in input a vtkPolyData corresponding to the surface and an optional distance matrix corresponding to the distance between the points of the surface in a metric space. 

If the distance matrix is not provided only information about the surface is computed, otherwise they are also computed for the metric space and a ratio or a difference between information about the surface and the metric space is computed to allow to compare them easily.

### CODE REVIEW

File: core/base/metricDistortion/MetricDistortion.h [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+        if(distanceMatrix.size() != 0)
+          ratioArea[i] = metricArea[i] / surfaceArea[i];
+      }
+    }
+
+    //-------------------------------------------------------------------------
+    // Utils
+    //-------------------------------------------------------------------------
+    double cosineLaw(double a, double b, double c) {
```
Comment (julien-tierny) [inline core/base/metricDistortion/MetricDistortion.h]: Please move all these low level geometric computations to `core/base/geometry/Geometry.h`

File: core/base/metricDistortion/MetricDistortion.h [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+      // Get the angle opposite to edge c
+      return std::acos((a * a + b * b - c * c) / (2.0 * a * b));
+    }
+
+    double l2Distance(std::vector<double> &p1, std::vector<double> &p2) {
```
Comment (julien-tierny) [inline core/base/metricDistortion/MetricDistortion.h]: This function is already implemented in `core/base/geometry/Geometry.h`.
Please use that implementation instead.

File: core/base/metricDistortion/MetricDistortion.h [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      double distance = 0.0;
+      for(unsigned int i = 0; i < p1.size(); ++i)
+        distance += std::pow(p1[i] - p2[i], 2);
+      return std::sqrt(distance);
+    }
+
+    double triangleArea3D(std::vector<double> &x1,
```
Comment (julien-tierny) [inline core/base/metricDistortion/MetricDistortion.h]: This function may be already implemented in `core/base/geometry/Geometry.h` (see `computeTriangleArea`)

File: core/base/metricDistortion/MetricDistortion.h [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        AB[i] = x2[i] - x1[i];
+        AC[i] = x3[i] - x1[i];
+      }
+      return std::sqrt(std::pow(AB[1] * AC[2] - AB[2] * AC[1], 2)
+                       + std::pow(AB[2] * AC[0] - AB[0] * AC[2], 2)
+                       + std::pow(AB[0] * AC[1] - AB[1] * AC[0], 2))
+             / 2.0;
+    }
+
+    double triangleAreaFromSides(double a, double b, double c) {
```
Comment (julien-tierny) [inline core/base/metricDistortion/MetricDistortion.h]: Please move this to `core/base/geometry/Geometry.h`

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+ * Specify the required input data type of each input port
+ *
+ * This method specifies the required input object data types of the
+ * filter by adding the vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE() key to
+ * the port information.
+ */
+int ttkMetricDistortion::FillInputPortInformation(int port,
+                                                  vtkInformation *info) {
+  if(port == 0) {
+    info->Set(vtkAlgorithm::INPUT_REQUIRED_DATA_TYPE(), "vtkPolyData");
```
Comment (julien-tierny) [inline core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp]: Side note for later: we may want to change this to `vtkPointSets` if we want to generalize to `vtkUnstructuredGrid` (to generalize the same kind of features to tetrahedral meshes).

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+  // --------------------------------------------------------------------------
+  // Get input object from input vector
+  // --------------------------------------------------------------------------
+  // Load Surface
+  vtkPolyData *inputSurface = vtkPolyData::GetData(inputVector[0]);
+  if(!inputSurface)
+    return 0;
+  auto noPoints = inputSurface->GetNumberOfPoints();
+  std::vector<std::vector<double>> surfacePoints(
```
Comment (julien-tierny) [inline core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp]: Please do not copy the input point data.
Please see `core/vtk/ttkGeometrySmoother` for a simple example of how to do it.
The core class will need to be adjusted (similarly to `core/base/scalarFieldSmoother`)

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    noPoints, std::vector<double>(3));
+  for(unsigned int i = 0; i < noPoints; ++i) {
+    double point[3];
+    inputSurface->GetPoints()->GetPoint(i, point);
+    for(unsigned int j = 0; j < 3; ++j)
+      surfacePoints[i][j] = point[j];
+  }
+  auto noCells = inputSurface->GetNumberOfCells();
```
Comment (julien-tierny) [inline core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp:160]: Same for the cells, see the GeometrySmoother example (use the Triangulation class)

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+  std::vector<std::vector<int>> surfaceCells(noCells);
+  for(unsigned int i = 0; i < noCells; ++i) {
+    auto noCellPoints = inputSurface->GetCell(i)->GetNumberOfPoints();
+    surfaceCells[i] = std::vector<int>(noCellPoints);
+    for(unsigned int j = 0; j < noCellPoints; ++j)
+      surfaceCells[i][j] = inputSurface->GetCell(i)->GetPointId(j);
+  }
+
+  // Detect boundary
```
Comment (julien-tierny) [inline core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp]: Same here, the triangulation class already implements many triangulation traversal features, including boundary access.
See `core/vtk/ttkScalarFieldCriticalPoints` and `core/base/scalarFieldCriticalPoints`.

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+      if(edgeBelonging[i][j] == 1) {
+        isPointBoundary[i] = true;
+        isPointBoundary[j] = true;
+      }
+    }
+  }
+
+  // Load Table
+  vtkTable *distanceMatrixVTK = vtkTable::GetData(inputVector[1]);
+  std::vector<std::vector<double>> distanceMatrix;
```
Comment (julien-tierny) [inline core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp]: Same here, no need to copy the input data.
Raw pointers can be extracted. Please see other TTK filters using vtkTables as an input.

File: paraview/xmls/MetricDistortion.xml [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+      <!-- OUTPUT PARAMETER WIDGETS -->
+
+      <!-- Create a UI group that contains all output parameter widgets (here only one) -->
+      
+      <!-- DEBUG -->
+      ${DEBUG_WIDGETS}
+
+      <!-- MENU CATEGORY -->
+      <Hints>
+        <ShowInMenu category="TTK - Misc" />
```
Comment (julien-tierny) [inline paraview/xmls/MetricDistortion.xml]: Please change for `TTK - Domain`.

File: standalone/MetricDistortion/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,23 @@
+cmake_minimum_required(VERSION 3.2)
```
Comment (julien-tierny) [inline standalone/MetricDistortion/CMakeLists.txt:1]: Thanks for the standalone :) (although that was optional)

**Review by julien-tierny — CHANGES_REQUESTED** [review]
hi @MatPont 
thanks a lot for the PR.
I have a number of comments dealing with features you implemented which are already implemented in TTK (in the Geometry and Triangulation classes). Thanks for using pre-existing functions (and moving a few non-pre-existing features there). I pointed several examples throughout TTK you can get inspiration from.
Thanks for addressing these!

File: core/base/metricDistortion/MetricDistortion.h (hunk 2 above)
Comment (MatPont) [reply core/base/metricDistortion/MetricDistortion.h]: Done in the next commit!

File: core/base/metricDistortion/MetricDistortion.h (hunk 3 above)
Comment (MatPont) [reply core/base/metricDistortion/MetricDistortion.h]: Done in the next commit!

File: core/base/metricDistortion/MetricDistortion.h (hunk 4 above)
Comment (MatPont) [reply core/base/metricDistortion/MetricDistortion.h]: Done in the next commit!

File: core/base/metricDistortion/MetricDistortion.h (hunk 1 above)
Comment (MatPont) [reply core/base/metricDistortion/MetricDistortion.h]: Done in the next commit!

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp (hunk 6 above)
Comment (MatPont) [reply core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp]: Done in the next commit.

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp (hunk 7 above)
Comment (MatPont) [reply core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp:160]: Done in the next commit.

File: paraview/xmls/MetricDistortion.xml (hunk 10 above)
Comment (MatPont) [reply paraview/xmls/MetricDistortion.xml]: Done in the next commit.

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp (hunk 5 above)
Comment (MatPont) [reply core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp]: Done in the next commit.

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp (hunk 8 above)
Comment (MatPont) [reply core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp]: Done in the next commit.

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp (hunk 9 above)
Comment (MatPont) [reply core/vtk/ttkMetricDistortion/ttkMetricDistortion.cpp]: Done in the next commit.

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.h [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+class TTKMETRICDISTORTION_EXPORT ttkMetricDistortion
+  : public ttkAlgorithm // we inherit from the generic ttkAlgorithm class
+  ,
+    protected ttk::MetricDistortion // and we inherit from the base class
+{
+private:
+  // Filled by the algorithm
+  // Area
+  std::vector<double> surfaceArea, metricArea, ratioArea;
```
Comment (julien-tierny) [inline core/vtk/ttkMetricDistortion/ttkMetricDistortion.h]: by convention in ttk, member variables have a "_" suffix, to distinguish them better.
for instance `surfaceArea` -› `surfaceArea_`

Comment (julien-tierny) [reply core/vtk/ttkMetricDistortion/ttkMetricDistortion.h]: Could you do that for all these vectors really quick?

**Review by julien-tierny — CHANGES_REQUESTED** [review]
that's close to perfect :)
I have one last tiny comment. if you could address that really quick, that'd be awesome :)

File: core/vtk/ttkMetricDistortion/ttkMetricDistortion.h (hunk 12 above)
Comment (MatPont) [reply core/vtk/ttkMetricDistortion/ttkMetricDistortion.h]: Indeed! Thanks, done in the next commit.

**Comment by julien-tierny** [comment]
ok, let's go!

**Pull request merged**